### ISSUE TITLE
SoftmaxLayer example

### ISSUE THREAD

**Issue by JunMa11** [cvxgrp/cvxpylayers#29]
Dear cvxpylayers team,

Thanks for sharing the great work.

I want to use the `softmax` layers in pytorch just like the ReLU layer in `example` folder.
https://github.com/cvxgrp/cvxpylayers/blob/master/examples/torch/ReLU%20Layers.ipynb

The expected input is a tensor with `shape=[batch, channel, x, y, z]` and output is the softmax results along the channel dim.
Just like `torch.nn.functional.softmax(input, dim=None, _stacklevel=3, dtype=None)`
https://pytorch.org/docs/stable/nn.functional.html#torch.nn.functional.softmax.

However, I have trouble to implement it.

```python
class SoftmaxOptiLayer(torch.nn.Module):
    """
    Softmax as an optimization layer
    An alternative of torch.nn.functional.softmax(input, dim=None, _stacklevel=3, dtype=None)

    Parameters:
    input: logits, shape=[batch, channel, x, y, z]
    dim: A dimension along which softmax will be computed.

    """
    def __init__(self, dim=None):
        super(SoftmaxOptiLayer, self).__init__()
        self.dim = dim

    def forward(self, input):
        # when x is batched, repeat W and b 
        _x = cp.Parameter(input)
        _y = cp.Variable(input)
        obj = cp.Minimize(-_x*_y - cp.sum(cp.entr(_y)))
        cons = ? #  how to compute the inner product along the specified dim?
        prob = cp.Problem(obj, cons)
        layer = CvxpyLayer(prob, parameters=[_x], variables=[_y])

        return layer(input)
```

In addition to ReLU, softmax is also a commonly used layer in CNNs.

**Would it be possible for you to add a `SoftmaxLayer` in the [examples/torch](https://github.com/cvxgrp/cvxpylayers/tree/master/examples/torch) folder?**

Looking forward to your reply.

Best regards,
Jun

**Comment by bamos** [comment]
Hi -- thanks for reaching out and trying our code! On the CVXPY side I recommend formulating this so that the inputs to the cvxpylayer are always shaped as `[batch, n]`  and doing the other reshaping in other PyTorch code. As a starting point you can use the variational softmax implementation we have here: https://locuslab.github.io/2019-10-28-cvxpylayers/

I'll close this for now but let us know if there's anything else!

**Comment by JunMa11** [comment]
Dear @bamos ,

Thanks for your quick reply.
I have read the variational softmax implementation in the blog. In the blog, the input is a 1-dim vector.

>  I recommend formulating this so that the inputs to the cvxpylayer are always shaped as `[batch, n]`

I'm sorry that I don't get this point.

My input is `x.shape=[batch, class_channel, x, y, z]`. If I reshape it into `[batch, class_channel*x*y*z]`, the spatial information is broken. I still do not know how to deal with it in cvxpy.

Specifically, I do not know **how should I formulate the obj and cons with the 5D (shape=[batch, class_channel, x, y, z]) input tensor.**

```python
obj = cp.Minimize(-_x.T*_y - cp.sum(cp.entr(_y)))
cons = [np.ones(d, dtype=np.float32).T*_y == 1.]
```

Could you give me more guidance at your convenience?

Looking forward to your reply.

Kindest regards,
Jun
